### PR TITLE
editor: move completion functions

### PR DESCRIPTION
**Pull request by flatcap**
**NOTE**: this _is_ working, but I need to do some more testing.

Deep within libenter there are lots of dependencies caused by auto-completion and history.

Users call `mw_get_field()` with arguments like `COMP_ALIAS` which determine which completion function to use and where to save the history.
Worse still is the fact that the alias code is in libenter.
and browser code and notmuch code and pattern code...

This PR creates a basic Completion API that passes in a function to do the work.
It should help with: #3955  Neomutt completion API

All the functions on the right are now in more appropriate places.
The abstraction isn't perfect, but it's a step forward.

Source: [graphviz](https://flatcap.org/mutt/keys/op_editor_complete.gv), [svg](https://flatcap.org/mutt/keys/op_editor_complete.svg)

### CODE REVIEW

**Comment by Amudtogal** [comment]
Hey flatcap,
this is amazing, thank you for the great refactor.
I will rebase my autocompletion branch on this, and this should speed things up.

File: complete/lib.h [hunk 1]
```diff
@@ -35,18 +35,25 @@
...
+#include <stddef.h>
 #include <stdbool.h>
 // IWYU pragma: begin_keep
+#include "compapi.h"
 #include "data.h"
 // IWYU pragma: end_keep
 
 struct Buffer;
 
+extern const struct CompleteOps CompleteCommandOps;
```
Comment (Amudtogal) [inline complete/lib.h:47]: why are these needed here?
if we use the `#include "compapi.h", are they not defined already (or is this an alias)?

Comment (flatcap) [reply complete/lib.h:47]: `compapi.h` defines the _interface_,
`CompleteCommandOps`, `CompleteLabelOps`, define _implementations_ which are used outside of the library.

We could do with better homes for these.

**Review by Amudtogal — APPROVED** [review]
Excellent, this is now so much clearer (at least in my head).

It's great that each user (alias, browser etc) manages their own completion function, without needing too much logic.

**Review by Amudtogal — COMMENTED** [review]
Excellent, this is now so much clearer (at least in my head).

It's great that each user (alias, browser etc) manages their own completion function, without needing too much logic.

**Comment by Amudtogal** [comment]
just one thing I spotted when compiling:
```
browser/complete.c:122:20: warning: implicit declaration of function ‘mutt_mb_is_shell_char’; did you mean ‘mutt_mb_is_pathname_char’? [-Wimplicit-function-declaration]
  122 |        (i > 0) && !mutt_mb_is_shell_char(wdata->state->wbuf[i - 1]); i--)
      |                    ^~~~~~~~~~~~~~~~~~~~~
      |                    mutt_mb_is_pathname_char

```

how would i best submit tiny patches like this?
is it best in a pull request, or assemble them along with the completion api branch?

File: complete/lib.h (hunk 1 above)
Comment (Amudtogal) [reply complete/lib.h:47]: I see. I wasn't aware that C implicitly declares functions as `extern`, and thus got confused by the mix in interface & implementation.

**Comment by flatcap** [comment]
> browser/complete.c:122:20: warning: implicit declaration of function ‘mutt_mb_is_shell_char’; did you mean ‘mutt_mb_is_pathname_char’? [-Wimplicit-function-declaration]

Hmm... I don't understand.

https://github.com/neomutt/neomutt/blob/b99bce75547070bfa43e491e80744ca1b409ccdb/browser/complete.c#L121-L124

should be fine, because the file includes `mutt/lib.h`

https://github.com/neomutt/neomutt/blob/b99bce75547070bfa43e491e80744ca1b409ccdb/browser/complete.c#L33

which includes `mbyte.h`

https://github.com/neomutt/neomutt/blob/b99bce75547070bfa43e491e80744ca1b409ccdb/mutt/lib.h#L87

which declares `mutt_mb_is_shell_char()`

https://github.com/neomutt/neomutt/blob/b99bce75547070bfa43e491e80744ca1b409ccdb/mutt/mbyte.h#L49

I can't repeat the error myself with gcc or clang.

> is it best in a pull request, or assemble them along with the completion api branch?

if you can figure out what's happened, push a commit to this branch.

> how would i best submit tiny patches like this?

Generally, feel free to create/add to a branch `[trivial]` in the neomutt repo.
I'll merge it into my next bundle of tidying.

**Comment by Amudtogal** [comment]
Fair enough, it is gone on my side as well, after `make clean && make`.
Maybe it was a lingering object file...
If it comes back up, I will contribute to `[trivial]`

**Comment by Amudtogal** [comment]
Ah found it:
further down in my dev branch, I am renaming `mutt_mb_is_shell_char` to `mutt_mb_is_pathname_char` (more verbose, and it was only used in one spot anyway).
This was causing the compilation problem.

I have rebased my dev branch on this pull request to get started.
